### DISCUSSION TITLE
Malt 1.0 Preview

### DISCUSSION THREAD

**Discussion by pragma37** [bnpr/Malt#231] [Announcements]
Malt 1.0 is finally feature-complete!!!

Now it's time for testing and feedback.

Release:
https://github.com/bnpr/Malt/releases/tag/Development-latest

[Material Examples](https://github.com/bnpr/Malt/discussions/94#discussioncomment-2117754)

### Render Nodes
The whole render pipeline can now be customized in node form:
![imagen](https://user-images.githubusercontent.com/36610999/152657342-3799f66e-688a-4bc7-a93a-a2f629ba9915.png)

### Pipeline Plugins
Existing pipelines can be customized through plugins.
Plugins can:
- Add new node libraries.
- Add new *Pipeline Parameters*.
- Add new *PipelineGraph types*.

Plugins can be installed globally in the *Addon Settings* or per world in the *World Panel*.

A basic example can be found in [Development/plugins](https://github.com/bnpr/Malt/tree/Development/plugins)

### AOVs in the UI
Custom IO can now be configured from the Blender user interface.
![imagen](https://user-images.githubusercontent.com/36610999/152657576-813b641c-4fed-41d2-aa68-86a6f2cc3180.png)

### Meta System
GLSL functions and structs can now declare default values, custom subtypes, and other properties (like visibility options) from source code:
```glsl
/*  META
    @position: subtype=Vector; default=POSITION;
    @normal: subtype=Normal; default=NORMAL;
    @light_group: default=1;
    @shadows: default=true;
    @self_shadows: default=true;
*/
vec3 diffuse_gradient_shading(vec3 position, vec3 normal, sampler1D gradient, int light_group, bool shadows, bool self_shadows)
{
    _LIT_SCENE_MACRO(diffuse_gradient_lit_surface(LS, gradient), light_group, shadows, self_shadows);
}
```
![imagen](https://user-images.githubusercontent.com/36610999/152657787-0e6c7d40-cf53-49dd-b4a3-e9160a7fb1d2.png)

### Shader API refactor
The whole shader API has been improved and cleaned up.
Functions have now good defaults, advanced functions marked as *internal* in the *Meta Properties* won't show in the nodes UI and the materials Inputs/Outputs have been redesigned
![imagen](https://user-images.githubusercontent.com/36610999/152657886-cfbb32f0-ff5e-4505-ac44-c222c8c7dacc.png)
> ⚠️ Please, note that this also means that existing shaders won't compile on 1.0.

### Improved Screen Shaders
Screen shaders now have access to shading and filters functions.
This allows doing deferred-style rendering and applying certain effects globally (like line art).

### Other improvements
There have been tons of other improvements. Including:
- Improved *Screen Shaders*, allowing deferred-style rendering, and applying certain effects globally (like line art).
- Mesh Material *Depth Offset*. Useful for NPR and other techniques like Parallax Occlusion Mapping.
- Improved *AMD compatibility*. This preview has been better tested on AMD GPUs and several bugs have been fixed.
- Smaller *UX improvements*, like auto-selecting Malt Node Trees based on object selection.

## About MacOS
Mac builds won't be provided anymore.
I don't have the hardware to test on, and its limited OpenGL support is holding Malt as a whole.
I hope to support it again at some point, but that's unlikely to happen anytime soon and there's no point in providing broken builds meanwhile.

## What's next?
Now it's time for [testing and feedback](https://github.com/bnpr/Malt/issues/new/choose)!
I won't be adding any new features until everything is stable, but there's still time for tweaking what's already there.
After that, backward compatibility will be guaranteed, there won't be more breaking changes unless there's a very good reason and the whole community agrees.

**Comment by Kolupsy** [comment]
Wow fantastic! This all sounds awesome. I am super excited for global libraries (that are now called Global Plugins right?) and default values for shader functions. My main question for default values would be how this works with multiple functions per file. Does every META 'comment' refer to the function declaration directly after it or are the defaults the same for every argument with the specified name in every function? I also like the auto-selecting of node trees. Thats very handy.

**Reply by pragma37** [reply]
>  I am super excited for global libraries (that are now called Global Plugins right?)

Yes, plugins are basically the updated versions of [this proposal](https://github.com/bnpr/Malt/discussions/148#discussioncomment-1478225).

> Does every META 'comment' refer to the function declaration directly after it or are the defaults the same for every argument with the specified name in every function?

The former. You can take a look at any of the built-in GLSL files for reference.
https://github.com/bnpr/Malt/blob/Development/Malt/Pipelines/NPR_Pipeline/Shaders/NPR_Pipeline/NPR_Shading.glsl#L18

>  I also like the auto-selecting of node trees. Thats very handy.

Thanks! I usually work with very few different node trees, but when I was working on the material samples having to manually switch between all the node trees was a pain. 😅

**Reply by Kolupsy** [reply]
Very nice! The defaults are working great but need a Pipeline refresh to work (I guess thats fine). Are there more properties to override other than `subtype` and `default`? For example something like `hidden = 1;` (hiding the default value) might be useful as well even if it is more cosmetic than functional.

**Reply by pragma37** [reply]
The default value is hidden when it's not a constant.
For example, when it's set to POSITION or NORMAL.
Aside from that, is there any good reason for hiding it in the UI?

From the top of my head, the other ones used at the moment are the *internal* meta property, for hiding internal functions in the nodes menu:
https://github.com/bnpr/Malt/blob/Development/Malt/Shaders/Filters/AO.glsl#L8
And *replace*, which can be used when a function name changes:
```glsl
/* META @meta: replace=old_function_name */
void new_function_name(){}
```

**Reply by Kolupsy** [reply]
![image](https://user-images.githubusercontent.com/88381426/153318466-758de66a-1963-42f7-9c2b-40865bce18c9.png)
Yes, now the vec4 multiply node is pretty bulky when spawned. But in general it can be useful to hide a default value (for example an UP vector which is by default (0,0,1) but its only displayed as a socket labeld 'Up')
How about min and max? Could you set those too? maybe have the default value even show a slider? If i knew what these meta properties represented I would be able to figure that out on my own but currently it is just blackbox to me

**Reply by pragma37** [reply]
Well, any default value that can't be evaluated by the glsl_eval function becomes hidden.
Since it doesn't support preprocessor macros, you could do something like this:
```glsl
#define FORCE_HIDDEN(value) value
/*META 
    @a: subtype=Vector; default=FORCE_HIDDEN(vec4(0));
    @b: subtype=Vector; default=FORCE_HIDDEN(vec4(0));
*/
vec4 vec4_add(vec4 a, vec4 b){ return a+b; }
```

> If i knew what these meta properties represented I would be able to figure that out on my own but currently it is just blackbox to me

Meta properties are just a way for the GLSLParser to attach metadata to functions and parameters.
They don't have any meaning by themselves, they are just arbitrary JSON entries.
The point is that you can later read those properties in the reflection data to do the actually useful stuff.
https://github.com/bnpr/Malt/search?l=Python&q=meta

**Reply by MrAdrianPl** [reply]
will there be or is it there any documentation about the nodes?

**Reply by pragma37** [reply]
@MrAdrianPl  I'm working on docs, but detailed documentation about every individual node seems unlikely.
For a general overview (although with some outdated bits) you can check [these videos](https://www.youtube.com/playlist?list=PLiN2BGdwwlLqbks8h5MohvH0Xd0Zql_Sg).
For how to use the nodes, the Material Examples are the best reference atm.

**Reply by MrAdrianPl** [reply]
@pragma37 sorry to bother you again but im wondering what would be a good solution for making a subsurf shader any tips or is there any existing material like that?

**Reply by pragma37** [reply]
@MrAdrianPl  No need to apologize, but it's best to ask about "how-to" questions in their own threads. 
https://github.com/bnpr/Malt/discussions/new?category=q-a
